presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateProjectOnValues.java

```diff
@@ -0,0 +1,165 @@
+
+        // exclude non-deterministic function
+        DeterminismEvaluator determinismEvaluator = new RowExpressionDeterminismEvaluator(functionAndTypeManager);
+        if (!projectRowExpressions.stream().allMatch(determinismEvaluator::isDeterministic)) {
+            return Result.empty();
+        }
+        if (!rows.stream().allMatch(row -> row.stream()
+                .allMatch(determinismEvaluator::isDeterministic))) {
+            return Result.empty();
+        }
+        ExpressionOptimizer expressionOptimizer = new RowExpressionOptimizer(metadata);
+
+        //rewrite ProjectNode assignment expressions
+        List<List<RowExpression>> rowExpressionsList = new ArrayList<>();
+        for (List<RowExpression> rowExpressions : rows) {
+            verify(rowExpressions.size() == valuesOutputVariables.size(), "Output variable does not match its value in ValuesNode");
+            Map<VariableReferenceExpression, RowExpression> mapping = Streams.zip(
+                            valuesOutputVariables.stream(),
+                            rowExpressions.stream(),
+                            SimpleImmutableEntry::new)
+                    .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+            List<RowExpression> rowExpressionsInProject = projectRowExpressions.stream()
+                    .map(expression -> inlineVariables(mapping, expression))
+                    .collect(toImmutableList());
+            rowExpressionsList.add(rowExpressionsInProject);
+        }
+
+        //Evaluate ProjectNode assignment expressions
+        List<List<Object>> rowValuesList = new ArrayList<>();
+        for (List<RowExpression> rowExpressionsInProject : rowExpressionsList) {
+            List<Object> rowValues = rowExpressionsInProject.stream()
+                    .map(rowExpression -> expressionOptimizer.optimize(
+                            rowExpression,
+                            OPTIMIZED,
+                            context.getSession().toConnectorSession(),
+                            variable -> variable))
+                    .collect(toImmutableList());
+            rowValuesList.add(rowValues);
+        }
+
+        //Form the ValuesNode transformed from ProjectNode
+        List<List<RowExpression>> rowExpressionsListInValuesNode = new ArrayList<>();
+        for (List<Object> rowValues : rowValuesList) {
+            List<RowExpression> rowExpressionsInValuesNode = Streams.zip(
+                    rowValues.stream(),
+                    projectOutputVariables.stream(),
+                    (elementValue, projectOutputVariable) ->
+                        (RowExpression) new ConstantExpression(elementValue, projectOutputVariable.getType())
+                ).collect(toImmutableList());
+            rowExpressionsListInValuesNode.add(rowExpressionsInValuesNode);
+        }
```

Comment:
Can't these three individual for loops be collapsed into a single for loop?

Reply:
Sorry for the misunderstanding.